jax/tree_util.py

```diff
@@ -253,3 +254,27 @@ def _namedtuple_node(t):
 NamedtupleNode = NodeType('namedtuple',
                           lambda xs: (tuple(xs), type(xs)),
                           lambda t, xs: t(*xs))
+
+
+class Partial(functools.partial):
+  """A version of functools.partial that works in pytrees.
+
+  Use it for partial function evaluation in a way that is compatibile with JAX's
+  transformations, e.g., ``Partial(func, *args, **kwargs)``.
+
+  (You need to explicitly opt-in to this behavior because we didn't want to give
+  functools.partial different semantics than normal function closures.)
+  """
+
+def _partial_to_iterable(partial_):
+  values = partial_.args + tuple(partial_.keywords.values())
```

Comment:
Does this depend on dict argument ordering being consistent (either when traversing the same un-mutated object, which IIRC CPython guarantees, or across CPython executions, which it doesn't)? If so, it might be best to treat them like we do dictionaries, i.e. require the keys to be sortable and to store things sorted.

In fact, if `partial_.args` and `partial_.keywords` are just a regular Python tuple and dict, respectively, we don't have to deconstruct them here. We could just do something like

```python
  values = (partial._args, partial_.keywords)
  spec = partial_.func
  return values, spec
```

(or compress that into a one-liner).

That way we can leverage the fact that pytrees work recursively and just reuse the existing tuple/dict handling code (i.e. no need for another round of sorting logic here).

That is, we could do

```python
register_pytree_node(Partial,
                     lambda partial_: ((partial_.args, partial_.keywords), partial_.func),
                     lambda func, xs: Partial(func, *xs[0], **xs[1]))
```

WDYT?

Reply:
Yes, I like leveraging the recursive nature of pytree, that's way better. Done.